### PR TITLE
[Ready for Review] Unapologetic Psych Buffs - Now with 80% less Mike Pence!

### PR DESCRIPTION
**Pull request by LordFowl**
This PR removes brain surgery's ability to instantly cure all traumas. Now traumas have been roughly divided into four categories of treatment:
1: Chakra therapy cures traumas that directly affect physical behavior, such as tourettes or paralysis. It is accomplished via the crystal therapy pod available to the psychiatrist.

2: Hypnotic therapy cures traumas that directly affect mental behavior, such as phobias or confusions. It is accomplished via the mesmetron pocketwatch available to the psychiatrist.

3: Isolation therapy cures traumas that cause hallucinations. It is accomplished via the metronome and facilitated by the isolation room available to the psychiatrist.

4: Surgical therapy cures traumas that do not fit into the above categories. It is accomplished via brain surgery.

Using the crystal therapy pod's neural scan function will provide the psychiatrist the brain damage the patient suffers from in numerical form. It falls upon the psychiatrist to determine how much of this brain damage is being caused by mental traumas the crystal therapy pod is equipped to deal with. Excessive cycles will cause increasing amounts of radiation damage.

The mesmetron pocketwatch requires the patient to believe in hypnosis. It then puts the patient into a slumber, where the psychiatrist can hypnotically suggest new behavior, provided the patient continues to believe in hypnosis. Each suggestion cures a single hypnotic trauma.
The isolation room is equipped with a metronome. When active, the metronome will have a chance every ticktock of the metronome to cure an isolation trauma, provided the patient is totally alone.

The psychiatry office has been expanded into a complete mental health ward. The only thing it is lacking is beds, as I believe recovering mental health patients still belong in the recovery ward.

The chaplain can also cure isolation and chakra traumas by dispelling the demons within with his null rod. Staring at the supermatter unprotected has been known to hypnotize. Electricity applied by any source directly to the head is also known to sometimes cure electroshock trauma.

image

New manual psych's spawn with:
image

- [x] Testing
- [x] Balance

### CODE REVIEW

**Comment by BurgerLUA** [comment]
I love how in depth this is but I think this is incredibly excessive in terms of difficulty and balance.

**Comment by OneOneThreeEight** [comment]
Uhm. How severe is the electroshock chair?

**Comment by LordFowl** [comment]
@OneOneThreeEight It deals either 5 burn on a failed roll or a fair chunk of Halloss. It is also of course based on the powernet, so this is with roundstart power availability.

**Comment by Arrow768** [comment]
There is already a feedback topic on the forum regarding this:
https://forums.aurorastation.org/viewtopic.php?p=94149#p94149

**Comment by LordFowl** [comment]
Will have to double-check on that merge update. Travis is spitting out some map errors.

File: code/game/objects/structures/therapy.dm [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+	return
+
+/obj/item/weapon/mesmetron
+	name = "mesmetron pocketwatch"
+	desc = "An elaborate pocketwatch, with captivating gold etching and an enchanting face. . ."
+	icon = 'icons/obj/items.dmi'
+	icon_state = "pocketwatch"
+	matter = list("glass" = 150, "gold" = 50)
+	w_class = 1
+	var/mob/living/carbon/human/thrall
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: Don't keep hardrefs. Use a weakref preferably.

File: code/game/objects/structures/therapy.dm [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+	icon = 'icons/obj/Cryogenic2.dmi'
+	icon_state = "syndipod_0"
+	density = 1
+	anchored = 1
+
+	use_power = 1
+	idle_power_usage = 60
+	active_power_usage = 10000
+
+	var/mob/living/carbon/occupant
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: Same deal here. The human cannot effectively clean itself up from here, so this should be a weakref.

File: code/game/objects/structures/therapy.dm [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+	if (G.affecting.abiotic())
+		user << "<span class='warning'>Subject cannot have abiotic items on.</span>"
+		return
+	if(locked)
+		user << "<span class='warning'>The pod is locked.</span>"
+		return
+
+	if(istype(G, /obj/item/weapon/grab))
+
+		var/mob/living/L = G:affecting
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: Unnecessary `:`. G is already of the proper type.

File: code/game/objects/structures/therapy.dm [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+		src.occupant.client.eye = src.occupant.client.mob
+		src.occupant.client.perspective = MOB_PERSPECTIVE
+	src.occupant.forceMove(get_turf(src))
+	src.occupant = null
+	update_use_power(1)
+	src.icon_state = "syndipod_0"
+	return
+
+/obj/machinery/chakrapod/attackby(obj/item/weapon/grab/G as obj, user as mob)
+	if ((!( istype(G, /obj/item/weapon/grab) ) || !( ismob(G.affecting) )))
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: You're checking the type of the grab here and below. Just do `if (!istype(G) || !ismob(G.affecting))` here and you can step one level to the front on the pyramid.

File: code/game/objects/structures/therapy.dm [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+		user << "<span class='warning'>The pod is already occupied!</span>"
+		return
+	if (G.affecting.abiotic())
+		user << "<span class='warning'>Subject cannot have abiotic items on.</span>"
+		return
+	if(locked)
+		user << "<span class='warning'>The pod is locked.</span>"
+		return
+
+	if(istype(G, /obj/item/weapon/grab))
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: See above, not necessary.

File: code/game/objects/structures/therapy.dm [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+
+		var/mob/living/L = G:affecting
+		visible_message("[user] starts putting [G:affecting] into the pod bed.", 3)
+
+		if (do_mob(user, G:affecting, 30, needhand = 0))
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: `:`, unnecessary.

File: code/game/objects/structures/therapy.dm [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+		user << "<span class='notice'><B>The pod is already occupied!</B></span>"
+		return
+	if (M.abiotic())
+		user << "<span class='notice'><B>Subject cannot have abiotic items on.</B></span>"
+		return
+	if(locked)
+		user << "<span class='warning'>The pod is locked.</span>"
+		return
+
+	var/mob/living/L = O
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: M already exists. Why are you using L as well?

File: code/game/objects/structures/therapy.dm [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+		src.icon_state = "syndipod_1"
+		for(var/obj/Obj in src)
+			Obj.forceMove(get_turf(src))
+	src.add_fingerprint(user)
+	return
+
+/obj/machinery/chakrapod/ex_act(severity)
+	switch(severity)
+		if(1.0)
+			for(var/atom/movable/A as mob|obj in src)
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: `as mob|obj` is unnecessary here.

File: code/game/objects/structures/therapy.dm [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+				A.forceMove(get_turf(src))
+				ex_act(severity)
+			qdel(src)
+			return
+		if(2.0)
+			if (prob(50))
+				for(var/atom/movable/A as mob|obj in src)
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: `as mob|obj` is unnecessary here.

File: code/game/objects/structures/therapy.dm [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+					A.forceMove(get_turf(src))
+					ex_act(severity)
+				qdel(src)
+				return
+		if(3.0)
+			if (prob(25))
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: `as mob|obj` is unnecessary here.

File: code/game/objects/structures/therapy.dm [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+		M.forceMove(src)
+		src.occupant = M
+		update_use_power(2)
+		src.icon_state = "syndipod_1"
+		for(var/obj/Obj in src)
+			Obj.forceMove(get_turf(src))
+	src.add_fingerprint(user)
+	return
+
+/obj/machinery/chakrapod/ex_act(severity)
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: The only thing that changes here is the probability.

```DM
/obj/machinery/chakrapod/ex_act(severity)
	var/probability = 100
	if (severity == 2)
		probability = 50
	else if (severity == 3)
		probability = 25

	if (!prob(probability))
		return

	var/turf/T = get_turf(src)
	for (var/atom/movable/A in src)
		A.forceMove(T)
		A.ex_act(severity)

	qdel(src)
```

File: code/game/objects/structures/therapy.dm [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+				src.icon_state = "syndipod_scannerconsole-p"
+		else
+			icon_state = initial(icon_state)
+
+/obj/machinery/chakraconsole/Initialize()
+	. = ..()
+	for(var/obj/machinery/chakrapod/C in orange(1,src))
+		connected = C
+		break
+	connected.connected = src
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: Null exception possible.

```DM
if (connected)
	connected.connected = src
```

File: code/game/objects/structures/therapy.dm [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+		connected.connected = null
+	return ..()
+
+/obj/machinery/chakraconsole/power_change()
+	..()
+	if(stat & BROKEN)
+		icon_state = "syndipod_scannerconsole-p"
+	else
+		if (stat & NOPOWER)
+			spawn(rand(0, 15))
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: Addtimer.

File: code/modules/mob/living/carbon/brain/brain_item.dm [hunk 14]
```diff
@@ -188,8 +188,11 @@
...
 		qdel(trauma)
 
-/obj/item/organ/brain/proc/cure_all_traumas(cure_permanent = FALSE)
+/obj/item/organ/brain/proc/cure_all_traumas(cure_permanent = FALSE, cure_type = "")
 	for(var/X in traumas)
 		var/datum/brain_trauma/trauma = X
-		if(cure_permanent || !trauma.permanent)
-			qdel(trauma)
\ No newline at end of file
+		if(trauma.cure_type == cure_type || cure_type == "all")
```
Comment (skull132) [inline code/modules/mob/living/carbon/brain/brain_item.dm]: Use a define for "all" as well?

File: code/modules/mob/living/carbon/brain/brain_item.dm [hunk 15]
```diff
@@ -188,8 +188,11 @@
...
+			if(cure_permanent || !trauma.permanent)
+				qdel(trauma)
+				if(cure_type != "all")
```
Comment (skull132) [inline code/modules/mob/living/carbon/brain/brain_item.dm]: define.

File: code/game/objects/structures/therapy.dm [hunk 16]
```diff
@@ -0,0 +1,586 @@
...
+	var/response = input(user,"Input number of rotations","Therapy Pod","0")
+	var/alert = text2num(sanitize(response))
+	if(alert)
+
+		for(var/i=0;i<alert;i++)
+			sleep(100)
+			var/electroshock_trauma = 0
+			if(H && H == connected.occupant)
+				var/obj/item/organ/brain/sponge = H.internal_organs_by_name["brain"]
+				if(sponge && istype(sponge))
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: Use early returns please.

```dm
	var/obj/item/organ/brain/sponge = H.internal_organs_by_name["brain"]
	if (!istype(sponge) || !sponge.traumas.len)
		return // or break, whichever.
```

etc.

File: code/game/objects/structures/therapy.dm [hunk 17]
```diff
@@ -0,0 +1,586 @@
...
+					playsound(src.loc, 'sound/machines/ding.ogg', 50, 1)
+					return
+				user << "<span class='danger'>Error: Crystal depletion not detected. Terminating operation..</span>"
+				playsound(src, 'sound/machines/buzz-two.ogg', 50, 1)
+				visible_message("<span class='warning'>[connected] buzzes harshly.</span>", "<span class='warning'>You hear a sharp buzz.</span>")
+
+/obj/machinery/chakraconsole/proc/neural_check(var/mob/user, var/mob/living/carbon/human/H)
+	var/response = input(user,"Input number of rotations","Therapy Pod","0")
+	var/alert = text2num(sanitize(response))
+	if(alert)
```
Comment (skull132) [inline code/game/objects/structures/therapy.dm]: Early return.

```dm
	if (!alert)
		return
```

File: code/controllers/subsystems/processing/slow_process.dm [hunk 18]
```diff
@@ -0,0 +1,10 @@
+/var/datum/controller/subsystem/processing/slow/SSslow_process
+
+// Not really much to it, literally just a slower-ticking SSprocessing.
+
+/datum/controller/subsystem/processing/slow
+    name = "Processing (Slow)"
+    wait = 60
+
+/datum/controller/subsystem/processing/slow/New()
+    NEW_SS_GLOBAL(SSslow_process)
```
Comment (skull132) [inline code/controllers/subsystems/processing/slow_process.dm]: Unnecessary for one or few objects. Use the regular SSprocessing and use time delta calculations.

```dm
/some/obj
	var/time_last_ran = 0

/some/obj/proc/process()
	if (world.time - time_last_ran < 60)
		return

	time_last_ran = world.time
```

File: code/__defines/mobs.dm [hunk 19]
```diff
@@ -212,3 +212,8 @@
 #define BRAIN_TRAUMA_SPECIAL /datum/brain_trauma/special
 
 #define BRAIN_DAMAGE_INTEGRITY_MULTIPLIER 0.5
+
+#define CURE_CRYSTAL "crystal"
+#define CURE_SOLITUDE "solitude"
+#define CURE_HYPNOSIS "hypnosis"
+#define CURE_SURGERY "surgery"
```
Comment (skull132) [inline code/__defines/mobs.dm]: `#define CURE_ALL "all"`

**Pull request merged**